### ISSUE TITLE
Key stuck effect

### ISSUE THREAD

**Issue by DavidGZ** [freemint/freemint#87]
I've been investigating the "key stuck" effect that shows sometimes on some systems.
[[1]]( http://www.atari-forum.com/viewtopic.php?f=97&t=30966&p=308714), [[2]](http://www.atari-forum.com/viewtopic.php?f=100&t=31429&start=25#p346883), [[3]](http://www.atari-forum.com/viewtopic.php?f=92&t=33402#p341799)

This usually happens when moving the mouse while the system is doing some intensive work.
The effect appears when the interrupts are disable for too long and the interrupt handler for the keyboard ACIA doesn't read the stream from the keyboard controller fast enough so overrun errors arise and mouse packets get lost. When there is an overrun condition the kernel discards the byte coming from the keyboard and calls TOS keyboard error vector and this vector does nothing it just returns (in TOS and EmuTOS). In fact if I've read the code correctly none of the operating systems (TOS, EmuTOS, Magic or FreeMiNT) tries to deal with overrun errors.

Mouse multi packets in relative mode (the working mode for GEM) have this form [0xFX][delta x][delta y], the first packet indicates that is a mouse multi packet and which mouse button is pressed. If we loss for example the first packet the keyboard handler most probably takes the following delta packets as key presses which won't have a key break packet and the key stuck effect will show.

I tried to implement a solution for avoiding packet's misinterpretation filtering packets after an overrun condition until we get a new correct mouse packet header, but it didn't work because I'm not able to get the overrun condition flag always that one happens.

This is a correct keyboard packet stream, without overruns, when I just move the mouse around. First
byte is the packet value and second byte is the ACIA state register (0x83 no overrun and 0xA3 there is a overrun), as you can see we always have a [0xF8] (mouse packet not button pressed) followed by two delta values.

pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0x00 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x00 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0x00 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x01 0x83
pid   0 (MiNT): 0xF8 0x83
pid   0 (MiNT): 0xFF 0x83
pid   0 (MiNT): 0x01 0x83

This is a stream with overrun conditions:
 
pid  77 (toswin2): 0xF8 0x83
pid  80 (bash): 0x01 0x83
pid  80 (bash): 0x00 0x83
pid  80 (bash): 0xF8 0x83
pid  80 (bash): 0x02 0x83
pid  80 (bash): 0x00 0x83
pid  80 (bash): 0xF8 0x83
pid  80 (bash): 0x01 0x83
pid  80 (bash): 0xF8 0x83 <- overrun but not flagged
pid  80 (bash): 0x02 0x83 <- this delta will be taken as a key pressed 
pid  80 (bash): 0x00 0x83
pid  77 (toswin2): 0xF8 0x83
pid  80 (bash): 0x01 0x83
pid  80 (bash): 0x01 0x83
pid  80 (bash): 0xF8 0x83
pid  80 (bash): 0x00 0x83
pid  80 (bash): 0xF8 0x83 <- overrun but not flagged
pid  80 (bash): 0x01 0x83 <- this delta will be taken as a key pressed 
pid  80 (bash): 0x01 0xA3 <- overrun flagged
pid  80 (bash): 0xF8 0x83
pid  80 (bash): 0xF8 0xA3
pid  72 (AESSYS): 0x00 0x83
pid  80 (bash): 0xF8 0x83
pid  80 (bash): 0x01 0x83
pid  80 (bash): 0x01 0x83
pid  76 (ztask): 0xF8 0x83
pid  80 (bash): 0x02 0x83
pid  80 (bash): 0x02 0xA3
pid  80 (bash): 0xF8 0x83
pid  80 (bash): 0xF8 0xA3
pid  80 (bash): 0x01 0x83
pid  80 (bash): 0xF8 0x83
pid  80 (bash): 0x02 0x83
pid  80 (bash): 0x02 0xA3
pid  80 (bash): 0xF8 0x83
pid  80 (bash): 0x01 0x83
pid  80 (bash): 0x01 0x83
pid  80 (bash): 0xF8 0x83
pid  76 (ztask): 0x02 0x83
pid  76 (ztask): 0x00 0x83

As you can see in the log above, there has been some overruns and the ACIA didn't flagged them. I don't know why this is happening. May be I'm missing something but I don't know how to find a solution at the software level seeing the above log.

[Linux](https://github.com/torvalds/linux/blob/ba4dbdedd3edc2798659bcd8b1a184ea8bdd04dc/arch/m68k/atari/atakeyb.c#L117) tries to handle those overruns, but IMHO seeing the log above their code will fail avoiding the key stuck effect. But again perhaps I'm missing something it's strange that Linux people didn't test that code.

I've attached a kernel module to force keyboard overruns, after loading the module just write "echo 1 > /dev/keystuck" in the shell, during around 1 minute the interrupts will be disable during 2 ms intervals. 
[keystuck.zip](https://github.com/freemint/freemint/files/2114552/keystuck.zip)

Also I've attached a patch to log the keyboard packet stream. If you log with a 030 you won't need to load the module above because the overhead for the packet output will generate the overruns and the key stuck effect. The 060 is fast enough so the module above is necessary.
[keystuck.txt](https://github.com/freemint/freemint/files/2114648/keystuck.txt)

**Comment by mikrosk** [comment]
But how would you explain the first post's "fix" was to tweak some CT60 stuff?

**Comment by DavidGZ** [comment]
The fix was to disable DMA in the video dialog in the CT60 config CPX, my guess is that this change in the CT60 TOS (probably version 2.0x) makes the system more responsive when some interrupts are requested.

**Comment by mikrosk** [comment]
I guess the ultimate proof that it is really some kind of HW issue/corner case would be to reproduce it under TOS. I think in one of the threads somebody mentioned being able to replicate the behaviour in MagiC, too.

**Comment by DavidGZ** [comment]
With only FireTOS is easily replicable specially with latest versions. Also I've compiled the kernel with -DNO_AKP_KEYBOARD flag, this disables the use of the kernel's keyboard routines, so it uses the ones from the underlying TOS, then I've tested that kernel with TOS and EmuTOS on a Falcon and the key stuck problems persisted.

**Comment by DavidGZ** [comment]
>This is far more pronounced on firetos and hence I get this stuck key repeat thing fairly often.

Yes, mine too specially with latest FireTOS versions, the system is unusable. For now the only solution is to use a USB mouse. With EmuTOS I have seen it too but it's not so bad, at least not worse than with my Falcon. I think at some point Didier added a task to the RTOS part of FireTOS that keeps interrupts disabled quite long or increased this interrupt disabled time in some of the existing tasks.

**Comment by DavidGZ** [comment]
I've just tested with Eiffel and the streams with overruns is very similar to the one with the original keyboard.

**Comment by jvaltane** [comment]
I also tried sometime ago Falcon with STe:s keyboad. It did not fix the problem. Second link at first comment is about my experiences. Basically got "key stuck" while networking, especially with heavy traffic.

Usually hitting esc-key is enough to stop the situation. Most cases after esc-key trick I was even able to use mouse without problems. How ever I have to test if this is true while transfer big file trough network or just with browsing web. I also had similar problems at some point with netusbee while using it for networking, but last time I tested it worked just fine.

So looks like FreeMiNTs network drivers causes with my set up these "key stucks". For those who has not read the source code, most network drivers uses interrupts quite heavily.

**Comment by DavidGZ** [comment]
I don't think there is solution for this problem from the software side, only trying to keep interrupts disable as brief as possible when coding drivers and even so sometimes won't be enough, as an example is the NetUSBee USB driver for the 030. Also we can try to facilitate (supporting with software) the use of mouse alternatives to the mouse connector in the keyboard.

**Comment by DavidGZ** [comment]
This is a very clever solution by Christian Zietz to avoid this issue when writing drivers:
https://github.com/freemint/freemint/commit/a197e1be69fde2b06bcd51a881e2ba214ba2f878#diff-df40c9d52c98042a61a69e10db790983

**Comment by mikrosk** [comment]
I guess this is still relevant and hasn't been fixed?

**Comment by infusixllc** [comment]
I don't see any related commits but I would imagine it still needs to be fixed.   You can pretty easily trigger this behavior by running freemint under firetos on a firebee without an SD card plugged in.  Ideally no matter what happened to the system we wouldn't get the stuck key (or mouse cursor flying) effect.

**Comment by mikrosk** [comment]
Can you test the latest build to see whether it is still there? As you say, most likely it is.

**Comment by DavidGZ** [comment]
This is a difficult subject because there is not a specific thing to be fixed with a specific culprit. In my opinion is a mix of a "weak" hardware/protocol design (ikbd protocol) with written software which ignores that design fragility, mostly drivers which keep interrupts disable for too long.
I don't think that we can fix this writing a new ikbd driver in the kernel, at least after thinking about it I couldn't come up with anything.
I think the only solution is to identify which drivers are abusing of keeping the interrupts disable for too long and add the brilliant solution implemented by @czietz for the isp116x USB host controller drivers.
So from the kernel point of view we can fix things if the culprit drivers are in FreeMiNT's repository, I don't know if this is the case for the network driver pointed above by @jvaltane. But in the case of FireTOS described by @infusixllc I'm afraid there is nothing we can do.

**Comment by mikrosk** [comment]
> I think the only solution is to identify which drivers are abusing of keeping the interrupts disable for too long and add the brilliant solution implemented by @czietz for the isp116x USB host controller drivers.

Yes, this is what I meant by "fixed", from your thorough investigation @DavidGZ I have come to the same conclusion, unfortunately.
